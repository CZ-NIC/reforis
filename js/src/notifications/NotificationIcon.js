/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

NotificationIcon.propTypes = {
    severity: propTypes.string.isRequired,
    className: propTypes.string,
};

export default function NotificationIcon({severity, className}) {
    let iconName = null;
    switch (severity) {
        case 'news':
            iconName = 'newspaper';
            break;
        case 'restart':
            iconName = 'power-off';
            break;
        case 'error':
            iconName = 'exclamation-circle';
            break;
        case 'update':
            iconName = 'sync';
            break;
        default:
    }

    return <i className={`fa fa-${iconName} ${className}`}/>;
}