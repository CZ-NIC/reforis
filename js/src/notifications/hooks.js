/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { useEffect, useState } from "react";

import API_URLs from "common/API";
import { useWSForisModule, useAPIGet, useAPIPost } from "foris";

const WS_MODULE = "router_notifications";

export default function useNotifications(ws) {
    const [notifications, setNotifications] = useState([]);
    const [getState, get] = useAPIGet(API_URLs.notifications);

    useEffect(() => {
        get();
    }, [get]);
    useEffect(() => {
        if (getState.data) {
            const nonDisplayedNotifications = getState.data.notifications
                .filter((notification) => !notification.displayed);
            setNotifications(nonDisplayedNotifications);
        }
    }, [getState]);

    const [WSCreateData] = useWSForisModule(ws, WS_MODULE, "create");
    const [WSMarkAsDisplayedData] = useWSForisModule(ws, WS_MODULE, "mark_as_displayed");
    useEffect(() => {
        if (WSCreateData || WSMarkAsDisplayedData) get();
    }, [WSCreateData, WSMarkAsDisplayedData, get]);

    const [, post] = useAPIPost(API_URLs.notifications);

    function dismiss(notificationId) {
        post({ ids: [notificationId] });
        setNotifications((currentNotifications) => currentNotifications.filter(
            (notification) => notification.id !== notificationId,
        ));
    }

    function dismissAll() {
        post({ ids: notifications.map((notification) => notification.id) });
        setNotifications([]);
    }

    return [notifications, dismiss, dismissAll];
}

const NEW_NOTIFICATION_ANIMATION_DURATION = 1; // Sec.

export function useNewNotification(ws) {
    const [newNotification, setNewNotification] = useState(false);
    useEffect(() => {
        if (newNotification) {
            setTimeout(
                () => setNewNotification(false),
                NEW_NOTIFICATION_ANIMATION_DURATION * 1000,
            );
        }
    }, [newNotification]);

    const [WSCreateData] = useWSForisModule(ws, WS_MODULE, "create");

    useEffect(() => {
        if (WSCreateData) setNewNotification(true);
    }, [WSCreateData]);

    return newNotification;
}
