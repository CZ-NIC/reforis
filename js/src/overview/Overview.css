.card {
    box-shadow: 0 0.75rem 1.5rem rgba(18,38,63,.03);
}

.card-header {
    background-color: white;
}

.card-body form table.table {
    margin-left: 0;
    max-width: none;
}

.card-body form button {
    max-width: none;
}