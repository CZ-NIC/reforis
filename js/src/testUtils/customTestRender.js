/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
