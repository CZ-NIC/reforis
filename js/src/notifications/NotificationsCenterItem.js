/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */
import React from 'react';
import NotificationIcon from "./NotificationIcon";
import {toLocaleDateString} from "./utils";

const BORDER_TYPES = {
    news: 'border-info',
    update: 'border-info',
    restart: 'border-danger',
    error: 'border-danger',
};

export default class NotificationsCenterItem extends React.Component {
    constructor(props) {
        super(props);
        this.myRef = React.createRef();
    }

    componentDidMount() {
        // Hack to scroll to needed item. ForisNotificationID is defined in the notification center template.
        if (window.ForisNotificationID === undefined)
            return;
        if (window.ForisNotificationID === this.props.notification.id)
            this.myRef.current.scrollIntoView({block: 'start', behavior: 'smooth'});
    }

    render() {
        const notification = this.props.notification;
        return <div ref={this.myRef} className={`card bg-light ${BORDER_TYPES[notification.severity]} sm-10`}>
            <div className="card-header">
                <NotificationIcon severity={notification.severity} className={'fa-2x'}/>
                <p className="text-muted">{toLocaleDateString(notification.created_at)}</p>
                <button
                    type="button"
                    className="close"
                    onClick={e => this.props.dismissHandler(notification.id)}
                >Ã—
                </button>
            </div>

            <div className="card-body">
                <p className="card-text">{notification.msg}</p>
            </div>
        </div>;
    }
}

