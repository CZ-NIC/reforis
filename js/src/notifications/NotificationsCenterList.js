/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useRef, useEffect} from 'react';
import propTypes from 'prop-types';

import {NOTIFICATION_PROP_TYPES, toLocaleDateString} from './utils';
import NotificationIcon from './NotificationIcon';

NotificationsCenterList.propTypes = {
    notifications: propTypes.arrayOf(NOTIFICATION_PROP_TYPES),
    dismiss: propTypes.func.isRequired
};

export default function NotificationsCenterList({notifications, dismiss}) {
    return notifications.map(
        notification => {
            return <NotificationsCenterItem
                key={notification.id}
                notification={notification}
                dismiss={() => dismiss(notification.id)}
            />
        }
    )
}

const BORDER_TYPES = {
    news: 'border-info',
    update: 'border-info',
    restart: 'border-danger',
    error: 'border-danger',
};

function NotificationsCenterItem({notification, dismiss}) {
    const myRef = useRef(null);

    function getIDFromURL() {
        const searchParams = new URLSearchParams(window.location.search);
        return searchParams.get('id') || null;
    }

    useEffect(() => {
        if (getIDFromURL() === notification.id)
            myRef.current.scrollIntoView({block: 'start', behavior: 'smooth'});
    }, []);


    return <div ref={myRef} className={`card bg-light ${BORDER_TYPES[notification.severity]} sm-10`}>
        <div className='card-header'>
            <NotificationIcon severity={notification.severity} className={'fa-2x'}/>
            <p className='text-muted'>{toLocaleDateString(notification.created_at)}</p>
            <button
                type='button'
                className='close'
                onClick={dismiss}
            >Ã—
            </button>
        </div>

        <div className='card-body'>
            <p className='card-text'>{notification.msg}</p>
        </div>
    </div>;
}
