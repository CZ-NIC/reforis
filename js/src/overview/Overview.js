/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { useAPIGet, withSpinnerOnSending, withErrorMessage } from "foris";
import API_URLs from "../common/API";
import DynamicFirewallCard from "./Cards/DynamicFirewallCard";
import DataCollectionCard from "./Cards/DataCollectionCard";
import AutomaticUpdatesCard from "./Cards/AutomaticUpdatesCard";
import NetmetrCard from "./Cards/NetmetrCard";
import ConnectionTest from "../connectionTest/ConnectionTest";
import OpenVPNClientsCard from "./Cards/OpenVPNClientsCard";
import Notifications from "../notifications/Notifications/Notifications";

import "./Overview.css";

Overview.propTypes = {
    ws: PropTypes.object.isRequired,
};

export default function Overview({ ws }) {
    const [packageList, getPackageList] = useAPIGet(API_URLs.packages);
    useEffect(() => {
        getPackageList();
    }, [getPackageList]);

    return (
        <OverviewWithErrorAndSpinner
            apiState={packageList.state}
            packages={packageList.data || {}}
            ws={ws}
        />
    );
}

function displayCard({ package_lists: packages }, cardName) {
    const enabledPackagesNames = [];
    packages
        .filter((item) => item.enabled)
        .map((item) => {
            enabledPackagesNames.push(item.name);
            item.options
                .filter((option) => option.enabled)
                .map((option) => {
                    enabledPackagesNames.push(option.name);
                    return null;
                });
            return null;
        });
    return enabledPackagesNames.includes(cardName);
}

OverviewCards.propTypes = {
    packages: PropTypes.object.isRequired,
    ws: PropTypes.object.isRequired,
};

function OverviewCards({ packages, ws }) {
    return (
        <>
            <h1>{_("Overview")}</h1>
            <div className="row row-cols-1 row-cols-lg-3 mt-4">
                <AutomaticUpdatesCard />
                <DataCollectionCard />
                <DynamicFirewallCard
                    activated={() => displayCard(packages, "dynfw")}
                />
                {displayCard(packages, "net_monitoring") ? (
                    <NetmetrCard />
                ) : null}
                <div className="col mb-4">
                    <div className="card h-100 user-select-none">
                        <div className="card-body">
                            <h6 className="text-uppercase text-muted mb-2">
                                {_("Connection Test")}
                                <a
                                    href="/reforis/network-settings/wan"
                                    className="text-secondary"
                                    title="Go to Connection Settings"
                                >
                                    <i className="fas fa-external-link-alt float-right" />
                                </a>
                            </h6>
                            <ConnectionTest ws={ws} type="wan" />
                        </div>
                    </div>
                </div>
                {displayCard(packages, "openvpn") ? (
                    <OpenVPNClientsCard />
                ) : null}
            </div>
            <Notifications ws={ws} />
        </>
    );
}

const OverviewWithErrorAndSpinner = withSpinnerOnSending(
    withErrorMessage(OverviewCards)
);
