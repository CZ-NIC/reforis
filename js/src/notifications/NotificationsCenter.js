/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


import React from "react";
import NotificationsWrapper from "./NotificationsWrapper";
import NotificationsCenterItem from "./NotificationsCenterItem";

class NotificationsCenter extends React.PureComponent {
    getDismissAllButton() {
        return <button
            type="button"
            id="btn-dismiss-all"
            className="btn btn-outline-danger float-right"
            onClick={this.props.dismissAllHandler}
        >
            {_('Dismiss all')}
        </button>
    }

    render() {

        return <div id='notifications-center' className="col-sm-10">
            <h3>{_('Settings')}</h3>
            <h3>{_('Notifications')}</h3>
            {
                this.props.notifications.length !== 0 ?
                    this.getDismissAllButton() :
                    <p className="text-muted text-center">{_('No notifications')}</p>
            }

            <NotificationsCenterItemsList
                notifications={this.props.notifications}

                dismissHandler={this.props.dismissHandler}
            />
        </div>
    }
}

function NotificationsCenterItemsList({notifications, dismissHandler}) {
    return notifications.map(
        notification => {
            return <NotificationsCenterItem
                key={notification.id}
                notification={notification}
                dismissHandler={dismissHandler}
            />
        }
    )
}

export default NotificationsWrapper(NotificationsCenter);