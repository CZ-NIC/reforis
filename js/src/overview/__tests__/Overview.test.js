/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import Overview, { displayCard } from "../Overview";
import { WebSockets } from "foris";
import mockAxios from "jest-mock-axios";
import {
    packageListsFixture,
    dataCollectionCardFixture,
    automaticUpdatesCardFixture,
    netmetrCardFixture,
    openVPNClientsCardFixture,
} from "./__fixtures__/overview";
import { notificationsFixture } from "../../notifications/__tests__/__fixtures__/notifications";
import { render, wait, getByText } from "foris/testUtils/customTestRender";

describe("Overview", () => {
    let overviewContainer;

    it("Snapshot of the whole page", async () => {
        const webSockets = new WebSockets();
        const { container } = render(<Overview ws={webSockets} />);

        mockAxios.mockResponse({ data: packageListsFixture });
        await wait(() => getByText(container, "Overview"));

        mockAxios.mockResponse({ data: automaticUpdatesCardFixture });
        await wait(() => getByText(container, "Automatic Updates"));

        mockAxios.mockResponse({ data: dataCollectionCardFixture });
        await wait(() => getByText(container, "Data Collection"));

        mockAxios.mockResponse({ data: netmetrCardFixture });
        await wait(() => getByText(container, "NetMetr"));

        mockAxios.mockResponse({ data: openVPNClientsCardFixture });
        await wait(() => getByText(container, "OpenVPN Clients"));

        mockAxios.mockResponse({ data: notificationsFixture });
        await wait(() => getByText(container, "Dismiss all"));
        overviewContainer = container;
        expect(overviewContainer).toMatchSnapshot();
    });

    it("Should display given cards", () => {
        expect(displayCard(packageListsFixture, "datacollect")).toBe(true);
    });

    it("Should display given options", () => {
        expect(displayCard(packageListsFixture, "dynfw")).toBe(true);
    });
});
