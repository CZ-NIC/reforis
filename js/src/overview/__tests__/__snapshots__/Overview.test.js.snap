// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview Snapshot of the whole page 1`] = `
<div>
  <h1>
    Overview
  </h1>
  <div
    class="row row-cols-1 row-cols-lg-3 mt-4"
  >
    <div
      class="col mb-4"
    >
      <div
        class="card user-select-none"
      >
        <div
          class="card-body"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col"
            >
              <h6
                class="text-uppercase text-muted mb-2 "
              >
                Automatic Updates
              </h6>
              <span
                class="status"
              >
                Activated
              </span>
            </div>
            <div
              class="col-auto"
            >
              <span
                class="h2 text-success"
              >
                <i
                  class="fas fa-check"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col mb-4"
    >
      <div
        class="card user-select-none"
      >
        <div
          class="card-body"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col"
            >
              <h6
                class="text-uppercase text-muted mb-2"
              >
                Threat Detection
              </h6>
              <span
                class="status"
              >
                Activated
              </span>
            </div>
            <div
              class="col-auto"
            >
              <span
                class="h2 text-success"
              >
                <i
                  class="fas fa-check"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col mb-4"
    >
      <div
        class="card user-select-none"
      >
        <div
          class="card-body"
        >
          <div
            class="row align-items-center"
          >
            <div
              class="col"
            >
              <h6
                class="text-uppercase text-muted mb-2"
              >
                Dynamic Firewall
              </h6>
              <span
                class="status"
              >
                Activated
              </span>
            </div>
            <div
              class="col-auto"
            >
              <span
                class="h2 text-success"
              >
                <i
                  class="fas fa-check"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col mb-4"
    >
      <div
        class="card h-100 user-select-none"
      >
        <div
          class="card-body"
        >
          <h6
            class="text-uppercase text-muted mb-2"
          >
            NetMetr
            <a
              class="text-secondary"
              href="/netmetr/speed-test"
              title="Go to NetMetr"
            >
              <i
                class="fas fa-chevron-right float-right"
              />
            </a>
          </h6>
          <form
            class="mb-4"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table table-borderless table-hover col-12"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      Download
                      <span
                        class="badge badge-secondary"
                        title="Megabit per second"
                      >
                        Mb/s
                      </span>
                    </th>
                    <td
                      class="text-right"
                    >
                      930
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Upload
                      <span
                        class="badge badge-secondary"
                        title="Megabit per second"
                      >
                        Mb/s
                      </span>
                    </th>
                    <td
                      class="text-right"
                    >
                      <span>
                        930
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Ping
                      <span
                        class="badge badge-secondary"
                        title="Millisecond"
                      >
                        ms
                      </span>
                    </th>
                    <td
                      class="text-right"
                    >
                      <span>
                        N/A
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      Link
                    </th>
                    <td
                      class="text-right"
                    >
                      <a
                        href="https://www.netmetr.cz/en/detail.html?c7a6f82d-5128-47f3-a3f9-d7e9d9af64bd"
                        rel="noopener noreferrer"
                        target="_blank"
                      >
                        Details
                        <sup>
                          <i
                            class="fas fa-external-link-alt fa-sm ml-1"
                          />
                        </sup>
                      </a>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </form>
          <p
            class="card-text text-to-bottom"
          >
            <small
              class="text-muted"
            >
              Performed
              Â 
              in 2 years
            </small>
          </p>
        </div>
      </div>
    </div>
    <div
      class="col mb-4"
    >
      <div
        class="card h-100 user-select-none"
      >
        <div
          class="card-body"
        >
          <h6
            class="text-uppercase text-muted mb-2"
          >
            Connection Test
          </h6>
          <div
            class=""
          >
            <form>
              <table
                class="table table-borderless table-hover col-sm-12"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      IPv4 connectivity
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-secondary"
                      >
                        <i
                          class="fas fa-minus"
                          title="Connection test not started"
                        />
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      IPv4 gateway connectivity
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-secondary"
                      >
                        <i
                          class="fas fa-minus"
                          title="Connection test not started"
                        />
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      IPv6 connectivity
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-secondary"
                      >
                        <i
                          class="fas fa-minus"
                          title="Connection test not started"
                        />
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      IPv6 gateway connectivity
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-secondary"
                      >
                        <i
                          class="fas fa-minus"
                          title="Connection test not started"
                        />
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      DNS
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-secondary"
                      >
                        <i
                          class="fas fa-minus"
                          title="Connection test not started"
                        />
                      </span>
                    </td>
                  </tr>
                  <tr>
                    <th
                      scope="row"
                    >
                      DNSSEC
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-secondary"
                      >
                        <i
                          class="fas fa-minus"
                          title="Connection test not started"
                        />
                      </span>
                    </td>
                  </tr>
                </tbody>
              </table>
              <div
                class="col-sm-12 col-lg-12 p-0 mb-0 text-right"
              >
                <button
                  class="btn btn-outline-primary mw-100 col-sm-12 col-md-3 col-lg-2"
                  type="submit"
                >
                  Test connection
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col mb-4"
    >
      <div
        class="card h-100 user-select-none"
      >
        <div
          class="card-body"
        >
          <h6
            class="text-uppercase text-muted mb-2"
          >
            OpenVPN Clients
            <a
              class="text-secondary"
              href="/openvpn/client-settings"
              title="Go to OpenVPN Client Settings"
            >
              <i
                class="fas fa-chevron-right float-right"
              />
            </a>
          </h6>
          <form>
            <div
              class="table-responsive"
            >
              <table
                class="table table-borderless table-hover col-sm-12"
              >
                <tbody>
                  <tr>
                    <th
                      scope="row"
                    >
                      <span>
                        turris
                      </span>
                    </th>
                    <td
                      class="text-right"
                    >
                      <span
                        class="text-success"
                      >
                        <i
                          class="fas fa-check"
                        />
                      </span>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
  <h2>
    Notifications
    <button
      class="btn btn-outline-danger float-right"
      id="btn-dismiss-all"
      type="button"
    >
      <i
        class="fas fa-trash mr-1"
      />
      Dismiss all
    </button>
  </h2>
  <div
    id="notifications-center"
  >
    <div
      class="card bg-light border-info"
    >
      <div
        class="card-header"
      >
        <i
          class="fa fa-newspaper fa-lg"
        />
        <p
          class="text-muted align-middle m-0"
        >
          February 1, 2000 12:00 AM
        </p>
        <button
          class="close "
          type="button"
        >
          Ã
        </button>
      </div>
      <div
        class="card-body"
      >
        <p
          class="card-text"
        >
          Notification message.
        </p>
      </div>
    </div>
    <div
      class="card bg-light border-danger"
    >
      <div
        class="card-header"
      >
        <i
          class="fa fa-exclamation-circle fa-lg"
        />
        <p
          class="text-muted align-middle m-0"
        >
          February 1, 2000 12:00 AM
        </p>
        <button
          class="close "
          type="button"
        >
          Ã
        </button>
      </div>
      <div
        class="card-body"
      >
        <p
          class="card-text"
        >
          Second notification message.
        </p>
      </div>
    </div>
    <div
      class="card bg-light border-danger"
    >
      <div
        class="card-header"
      >
        <i
          class="fa fa-power-off fa-lg"
        />
        <p
          class="text-muted align-middle m-0"
        >
          February 1, 2000 12:00 AM
        </p>
        <button
          class="close invisible"
          type="button"
        >
          Ã
        </button>
      </div>
      <div
        class="card-body"
      >
        <p
          class="card-text"
        >
          Third notification message.
        </p>
      </div>
      <button
        class="btn btn-danger"
        type="button"
      >
        Reboot
      </button>
    </div>
  </div>
</div>
`;
